### PR TITLE
Allow cache responses from local DNS servers

### PR DESCRIPTION
**Pull request by psapezhka**
Add option to allow cache responses from local DNS servers.

Fixes: #17679 #4480

### CODE REVIEW

File: man/resolved.conf.xml [hunk 1]
```diff
@@ -251,8 +251,16 @@
         request. Be aware that turning off caching comes at a performance penalty, which is particularly high
         when DNSSEC is used. If <literal>no-negative</literal>, only positive answers are cached.</para>
 
-        <para>Note that caching is turned off implicitly if the configured DNS server is on a host-local IP address
-        (such as 127.0.0.1 or ::1), in order to avoid duplicate local caching.</para></listitem>
+        <para>Note that caching is turned off by default for host-local DNS servers.
+            See <varname>CacheFromLocalhost=</varname> for details.</para></listitem>
```
Comment (yuwata) [inline man/resolved.conf.xml]: Spurious indentation.

File: src/resolve/resolved-dns-transaction.c [hunk 2]
```diff
@@ -682,12 +682,10 @@ static void dns_transaction_cache_answer(DnsTransaction *t) {
...
-        /* We never cache if this packet is from the local host, under
-         * the assumption that a locally running DNS server would
-         * cache this anyway, and probably knows better when to flush
-         * the cache then we could. */
-        if (!DNS_PACKET_SHALL_CACHE(t->received))
-                return;
+        /* Packet from localhost? */
+        if (!t->scope->manager->enable_caching_from_localhost &&
+            in_addr_is_localhost(t->received->family, &t->received->sender) != 0)
+            return;
```
Comment (yuwata) [inline src/resolve/resolved-dns-transaction.c]: Please use 8ch indentation for `return;`

File: src/resolve/resolved-manager.h [hunk 3]
```diff
@@ -39,6 +39,7 @@ struct Manager {
         DnssecMode dnssec_mode;
         DnsOverTlsMode dns_over_tls_mode;
         DnsCacheMode enable_cache;
+        bool enable_caching_from_localhost;
```
Comment (yuwata) [inline src/resolve/resolved-manager.h]: Not a matter, but `enable` prefix sounds redundant for boolean variable.

**Review by yuwata — COMMENTED** [review]
Several nitpicks.

File: man/resolved.conf.xml (hunk 1 above)
Comment (psapezhka) [reply man/resolved.conf.xml]: Done.

File: src/resolve/resolved-dns-transaction.c (hunk 2 above)
Comment (psapezhka) [reply src/resolve/resolved-dns-transaction.c]: Done.

File: src/resolve/resolved-manager.h (hunk 3 above)
Comment (psapezhka) [reply src/resolve/resolved-manager.h]: Agree, fixed.

File: src/resolve/resolved-manager.c [hunk 4]
```diff
@@ -590,6 +590,7 @@ int manager_new(Manager **ret) {
                 .dnssec_mode = DEFAULT_DNSSEC_MODE,
                 .dns_over_tls_mode = DEFAULT_DNS_OVER_TLS_MODE,
                 .enable_cache = DNS_CACHE_MODE_YES,
+                .cache_from_localhost = false,
```
Comment (poettering) [inline src/resolve/resolved-manager.c]: no need to list that here btw. if not specified things are initialized to 0 anyway.

Comment (psapezhka) [reply src/resolve/resolved-manager.c]: Fixed.

**Pull request merged**